compiler/rustc_ty_utils/src/abi.rs

```diff
@@ -371,6 +371,8 @@ fn fn_abi_new_uncached<'tcx>(
     let target = &cx.tcx.sess.target;
     let target_env_gnu_like = matches!(&target.env[..], "gnu" | "musl" | "uclibc");
     let win_x64_gnu = target.os == "windows" && target.arch == "x86_64" && target.env == "gnu";
+    let x86_sse2 =
+        target.arch == "x86" && (target.features.contains("sse2") || target.cpu == "pentium4");
```

Comment:
I expect you'd want to query the `sse2` target feature rather than checking for the `pentium4` cpu, since e.g. `penryn` also supports `sse2` but isn't the `pentium4`. https://rust.godbolt.org/z/j7oPE6Wrf

also, what if there's a soft-float target for an OS kernel where `sse2` would be disabled even if the cpu is `pentium4` due to the kernel not saving SSE registers?

Reply:
@GuentherVIII If you handle the aforementioned edge-cases appropriately, I believe we can get this approved. We want to only do this if the target's baseline features include SSE2.

Reply:
so, in other words, only if `SSE2` is implied by the target structs/json ignoring any `-C target-features` or `-C target-cpu` or other overrides.

Reply:
Does rust currently ask LLVM for the target features implied by the target cpu? I could only find the code that creates the list for `#[cfg(target_feature)]`, which takes command-line options into account.

I've changed the implementation to check that list instead of target.features and target.cpu. This has the advantage that the standard library can activate the NAN-returning test precisely when it passes, and deactivating SSE with -C target-features still works.

There's also a second commit introducing yet another target option to make this configurable in the target specification and enabling it in all "i686" targets except i686_unknown_uefi, which disables SSE. The standard

I do not think we should make the ABI dependant on `#[target_feature]` - unlike -C target-feature, the attribute cannot disable SSE. If someone disables SSE with -C target-feature and re-enables it for select functions with `#[target_feature]`, this shouldn't result in incompatible function ABIs with the same function type.

Reply:
the issue you'll run into with `-C target-feature=-sse` is that functions in your binary won't match the ABI of other code (such as `std`) with SSE2 enabled in the default target features

Reply:
There's talk about forbidding going below the baseline. If you want less-than-baseline features you need a different target.

https://rust-lang.zulipchat.com/#narrow/stream/136281-t-opsem/topic/Float.20ABI.20hell/near/394473707